cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/engine/sql/SQLEngineInputTypeAdapter.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.api.engine.sql;
+
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import io.cdap.cdap.api.data.batch.Input;
+import io.cdap.cdap.api.data.batch.InputTypeAdapter;
+
+import java.lang.reflect.Type;
+
+/**
+ * Adapter for serialization/deserialization of the {@link Input} class, with additional support for the child
+ * {@link SQLEngineInput} class
+ */
+public class SQLEngineInputTypeAdapter extends InputTypeAdapter {
```

Comment:
I think this should be a single type adapter otherwise it's chicken and egg problem: to decide which type adapter to use, GSON needs to know which class it's going to be. I advise to use more standard factory-like style with a enum that tells which class is that